### PR TITLE
feat(tracking): add tracking for some events

### PR DESCRIPTION
**Pull request by mchepelev**
Closes EWC-429

- Does the iofog.org or README.md documentation need to change?
  - [x] No

### CODE REVIEW

File: src/cli/config.js [hunk 1]
```diff
@@ -244,10 +246,14 @@ const _listConfigOptions = function () {
...
 };
 
-const _changeDevModeState = function (options) {
+const _changeDevModeState = async function (options) {
   const enableDevMode = AppHelper.validateBooleanCliOptions(options.on, options.off);
   config.set('Server:DevMode', enableDevMode);
-  logger.info('Dev mode state updated successfully.')
+  logger.info('Dev mode state updated successfully.');
+
+  //TODO add same for other config props
```
Comment (MaksimChepelev) [inline src/cli/config.js]: -todo +simple comment

File: src/helpers/app-helper.js [hunk 2]
```diff
@@ -302,6 +302,17 @@ function isEmpty(obj) {
   return true;
 }
 
+function isOnline() {
+  const daemon = require('../daemon');
+
+  let pid = daemon.status();
+  if (pid === 0) {
```
Comment (MaksimChepelev) [inline src/helpers/app-helper.js]: return pid === 0

File: src/services/agent-service.js [hunk 3]
```diff
@@ -491,8 +494,17 @@ async function _checkMicroservicesFogType(fog, fogTypeId, transaction) {
   }
 }
 
+async function postTracking(events, fog, transaction) {
+  await TrackingEventManager.bulkCreate(events, transaction);
+  return {};
```
Comment (MaksimChepelev) [inline src/services/agent-service.js]: check

File: src/routes/agent.js [hunk 4]
```diff
@@ -442,6 +442,27 @@ module.exports = [
...
+  },
+  {
+    method: 'post',
+    path: '/api/v3/agent/tracking',
+    middleware: async (req, res) => {
+      const successCode = constants.HTTP_CODE_SUCCESS;
+      const errorCodes = [
+        {
+          code: constants.HTTP_CODE_UNAUTHORIZED,
+          errors: [Errors.AuthenticationError]
```
Comment (Railag) [inline src/routes/agent.js]: 400 Bad Request + json schema validation

File: src/decorators/tracking-decorator.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+      return await f.apply(this, arguments);
+    }
+    try {
+      const fArgs = Array.prototype.slice.call(arguments); // TODO can be used later
+      const res = await f.apply(this, arguments);
+      const event = Tracking.buildEvent(eventType, fArgs, res, f.name);
+      await Tracking.processEvent(event, fArgs, res);
+      return res;
+    } catch (e) {
+      throw e;
```
Comment (MaksimChepelev) [inline src/decorators/tracking-decorator.js]: simplify

File: src/tracking/index.js [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+const crypto = require('crypto');
+
+const TrackingEventManager = require('../sequelize/managers/tracking-event-manager');
+const Transaction = require('sequelize/lib/transaction');
+
+const fakeTransactionObject = {fakeTransaction: true};
+
+const trackingUuid = getUniqueTrackingUuid();
+
+function buildEvent(eventType, args, res, functionName) {
```
Comment (MaksimChepelev) [inline src/tracking/index.js]: doc

Comment (MaksimChepelev) [reply src/tracking/index.js]: type, res, args, funcName

File: src/tracking/index.js [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+      await TrackingEventManager.create(event, fArgs[fArgs.length - 1]);
+    } else {
+      await TrackingEventManager.create(event, fakeTransactionObject);
+    }
+  } else {
+    //just send
+    try {
+      sendEvents([event]);
+    } catch (e) {
+      //ignore
```
Comment (Railag) [inline src/tracking/index.js]: log

File: src/tracking/index.js [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+  request.end();
+}
+
+function getUniqueTrackingUuid() {
+  let uuid;
+  try {
+    let allMacs = '';
+    const interfaces = os.networkInterfaces();
+    for (const i in interfaces) {
+      if (i.internal) {
```
Comment (MaksimChepelev) [inline src/tracking/index.js]: if !internal

File: src/cli/config.js [hunk 9]
```diff
@@ -18,6 +18,8 @@ const AppHelper = require('../helpers/app-helper');
 const ErrorMessages = require('../helpers/error-messages');
```
Comment (Railag) [inline src/cli/config.js]: add unit tests & update postman tests for tracking endpoint

File: src/jobs/send-tracking-job.js [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+const intervalMin = 5;
+
+async function sendTracking() {
+  const fakeTransactionObject = {fakeTransaction: true};
+  const events = await TrackingEventManager.popAll(fakeTransactionObject);
+  try {
+    Tracking.sendEvents(events);
+  } catch (e) {
+    //TODO log only in file. add after logging will fixed
```
Comment (alexandershpak) [inline src/jobs/send-tracking-job.js]: log?

**Pull request merged**